In [9]:
from datetime import datetime
import jalali_pandas
import numpy as np 
import pandas as pd

%load_ext autoreload
%autoreload 2

from datetime import datetime, timedelta
import MetaTrader5 as Mt5

from include.trade import Trade
from include.tick import Tick
from include.rates import Rates
from include.indicator_connector import Indicator

The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


In [ ]:
# def start_mt5(username, password, server, path):
#     # Ensure that all variables are the correct type
#     uname           = int(username) # Username must be an int
#     pword           = str(password) # Password must be a string
#     trading_server  = str(server) # Server must be a string
#     filepath        = str(path) # Filepath must be a string

#     # Attempt to start MT5
#     if Mt5.initialize(login=uname, password=pword, server=trading_server, path=filepath):
#         # Login to MT5
#         if Mt5.login(login=uname, password=pword, server=trading_server):
#             return True
#         else:
#             print("Login Fail")
#             quit()
#             return PermissionError
#     else:
#         print("MT5 Initialization Failed")
#         quit()
#         return ConnectionAbortedError
    
# start_mt5(  
#         username  = 5014120852       ,
#         password  = 'blqlyf5p'       ,
#         server    = 'MetaQuotes-Demo',
#         path      = rf'C:\Users\s.tajedin\AppData\Roaming\MetaTrader 5\terminal64.exe'
#     )


In [3]:
Mt5.initialize()

True

## Get 5 Min Data

In [ ]:
from_data =datetime.now()
to_date =datetime(2023,1,10)
Number_candle = 15
rates = Mt5.copy_rates_from('XAUUSD',Mt5.TIMEFRAME_M5, from_data,Number_candle)

df_rates = pd.DataFrame(rates)
df_rates['time'] =pd.to_datetime(df_rates['time'], unit="s")#.jalali.to_jalali()

df_rates

In [ ]:
trade = Trade(
            expert_name           = 'TEST_01',      # Expert name
            version               = 0.1,            # Expert Version
            symbol                = 'XAUUSD',       # Symbol
            magic_number          = 1245,           # Magic Number
            lot                   = 1523.0,           # Volume
            stop_loss             = 20,             # It calls a functions that tries to close the deal, it is the stop you want.
            emergency_stop_loss   = 100,            # It set stop on chart as a protection if something went wrong with the stop_loss.
            take_profit           = 20,             # The same of stop_loss but for profit.
            emergency_take_profit = 100,            # The same of emergency_stop_loss but for gain
            start_time            = '9:15',         # It is the hour and minutes that the expert advisor are able to start to run.
            finishing_time        = '17:30',        # No new position can be opened after that time.
            ending_time           = '17:50',        # If there is a position opened it will be closed.
            fee                   = 0.0,            # fee
        )

In [ ]:
trade.statistics()

## 1 Step back Candle

In [ ]:
rates = Rates(trade.symbol, Mt5.TIMEFRAME_M30, 1, 1)

In [ ]:
print(rates.close)
print(rates.open)
print(rates.high)
print(rates.low)
print(rates.real_volume)
print(rates.spread)
print(rates.tick_volume)
print(rates.time)

### Tick
> #### current price

In [ ]:
tick = Tick(trade.symbol)

In [ ]:
print(tick.ask)
print(tick.bid)
print(tick.last) # price for last Trade
print(tick.volume)
print(tick.volume_real)
print(tick.flags)

In [ ]:
tick.time_msc

In [ ]:
print(tick.last)
print(rates.open)

In [4]:
trade = Trade(
            expert_name           = 'TEST_01',      # Expert name
            version               = 0.1,            # Expert Version
            symbol                = 'XAUUSD',       # Symbol
            magic_number          = 1245,           # Magic Number
            lot                   = 0.1,            # Volume
            stop_loss             = 250,             # It calls a functions that tries to close the deal, it is the stop you want.
            emergency_stop_loss   = 100,            # It set stop on chart as a protection if something went wrong with the stop_loss.
            take_profit           = 25,             # The same of stop_loss but for profit.
            emergency_take_profit = 250,            # The same of emergency_stop_loss but for gain
            start_time            = '9:15',         # It is the hour and minutes that the expert advisor are able to start to run.
            finishing_time        = '17:30',        # No new position can be opened after that time.
            ending_time           = '17:50',        # If there is a position opened it will be closed.
            fee                   = 0.5,            # fee
        )



Initializing the basics.
You are running the TEST_01 expert advisor, version 0.1, on symbol XAUUSD.
Initialization successfully completed.

Summary:
ExpertAdvisor name:              TEST_01
ExpertAdvisor version:           0.1
Running on symbol:               XAUUSD
MagicNumber:                     1245
Number of lot(s):                0.1
StopLoss:                        250
TakeProfit:                      25
Emergency StopLoss:              100
Emergency TakeProfit:            250
Start trading time:              9:15
Finishing trading time:          17:30
Closing position after:          17:50
Average fee per trading:         0.5
StopLoss & TakeProfit Steps:     1.0

Running



In [18]:
time = 0
while True:

    tick = Tick(trade.symbol)
    rates =Rates(trade.symbol, Mt5.TIMEFRAME_M5, 1, 1)

    if tick.time_msc != time:
        buy  = (tick.last > rates.open)
        sell = (tick.last < rates.open)

        trade.open_position(buy, sell, 'Example Advisor')

    time = tick.time_msc

    if trade.days_end():
        trade.close_position('End of the trading day reached.')
        break

print('Finishing the program.')
print('Program finished.')

Unexpected exception formatting exception. Falling back to standard exception
Order sent: XAUUSD, 0.1 lot(s), at 1960.53.
--------------------------------------------------
Something went wrong while retrieving ret_code, error: 10018
--------------------------------------------------
Error Discription :
Market is closed
--------------------------------------------------
Statistics :
Total of deals: 0, 0 gain, 0 loss.
Balance: 0.0, fee: 0.0, final balance: 0.0.
--------------------------------------------------


Traceback (most recent call last):
  File "c:\Users\s.tajedin\Miniconda3\envs\metatrader\lib\site-packages\IPython\core\interactiveshell.py", line 1993, in showtraceback
    stb = self.InteractiveTB.structured_traceback(
  File "c:\Users\s.tajedin\Miniconda3\envs\metatrader\lib\site-packages\IPython\core\ultratb.py", line 1118, in structured_traceback
    return FormattedTB.structured_traceback(
  File "c:\Users\s.tajedin\Miniconda3\envs\metatrader\lib\site-packages\IPython\core\ultratb.py", line 1012, in structured_traceback
    return VerboseTB.structured_traceback(
  File "c:\Users\s.tajedin\Miniconda3\envs\metatrader\lib\site-packages\IPython\core\ultratb.py", line 865, in structured_traceback
    formatted_exception = self.format_exception_as_a_whole(etype, evalue, etb, number_of_lines_of_context,
  File "c:\Users\s.tajedin\Miniconda3\envs\metatrader\lib\site-packages\IPython\core\ultratb.py", line 818, in format_exception_as_a_whole
    frames.append(self.format_record(r))
  File

SystemExit: Exiting the code with sys.exit()!

### چون در ایران معاملات الگوریتمیک ممنوع می باشد

> 10026 : Autotrading disabled by server \
> 10027 : Autotrading disabled by client terminal \
> 10014 : Invalid volume in the request \
> 10018 : Market is closed

In [ ]:
dur_   = timedelta(minutes=120)
now_  =datetime.now()
from_ = now_ - dur_
print(now_)
print(from_)

Mt5.history_deals_get(from_,now_,'XAUUSD')

In [ ]:
trade.statistics()

In [ ]:
from_date=datetime(2020,1,1)
to_date=datetime.now()
deals=Mt5.history_deals_get(from_date, to_date, group="*xauusd*")

In [ ]:
if deals==None:
    print("No deals with group=\"*USD*\", error code={}".format(Mt5.last_error()))
elif len(deals)> 0:
    print("history_deals_get({}, {}, group=\"*GBP*\")={}".format(from_date,to_date,len(deals)))
elif len(deals)< 0:
    print("history_deals_get({}, {}, group=\"*GBP*\")={}".format(from_date,to_date,len(deals)))
elif len(deals)+ 0:
    print("history_deals_get({}, {}, group=\"*GBP*\")={}".format(from_date,to_date,len(deals)))
    